### PR TITLE
fix unmounted setState warnings

### LINKED ISSUE #1449: Check didMount in setStatus to avoid calling setState after Formik unmounts
I'm currently using Formik in my project (loving it), but ran into an issue recently.

## Problem

All of the onSubmit handlers in my app return a promise. I wrap each handler using a common HOF like so, to handle the submitting and status state in a consistent way when the promise resolves/rejects:

```
export const handleSubmit = (onSubmit) => (data, {
	setStatus,
	setSubmitting,
}) => {
	setStatus(null);
	return onSubmit(data)
		.then((result) => setStatus(result))
		.catch((error) => {
			const { message: id, values } = error;
			setStatus({ type: 'error', message: { id, values } });
		})
		.finally(() => setSubmitting(false));
};
```

Some of those onSubmit handlers result in a route change on success, which unmounts the Formik instance. I'm using the Next.js router to handle the route change, so in this case I return the promise returned by the router's push/replace methods. This is a lot of detail to say: The `then`, `catch`, and `finally` handlers above end up firing _after_ the route change — that is, _after_ the Formik instance has unmounted.

This is actually not a problem for the `finally` handler because `setSubmitting` (very smartly) guards against calling setState after the component has unmounted — but not so for setStatus 😞 (nor for any of the other setters available in the FormikBag).

## Proposed Solution

```
setStatus = (status?: any) => {
    if (this.didMount) {
        this.setState({ status });
    }
};
```

### PR DESCRIPTION
**Pull request by armandooj**
Resolves #1449.

### CODE REVIEW

File: src/Formik.tsx [hunk 1]
```diff
@@ -120,27 +120,35 @@ export class Formik<Values = FormikValues> extends React.Component<
   }
 
   setErrors = (errors: FormikErrors<Values>) => {
-    this.setState({ errors });
+    if (this.didMount) {
```
Comment (KITSEPAMIgorSergiienko) [inline src/Formik.tsx]: maybe create something like this to not repeat code every time?
```js
safeSetState = (...args) => {
  if (this.didMount) {
    this.setState(...args)
  }
}

File: src/Formik.tsx [hunk 2]
```diff
@@ -506,26 +514,30 @@ export class Formik<Values = FormikValues> extends React.Component<
...
   setFieldError = (field: string, message: string | undefined) => {
-    // Set form field by name
-    this.setState(prevState => ({
-      ...prevState,
-      errors: setIn(prevState.errors, field, message),
-    }));
+    if (this.didMount) {
+      // Set form field by name
+      this.setState(prevState => ({
+        ...prevState,
```
Comment (KITSEPAMIgorSergiienko) [inline src/Formik.tsx]: `...prevState` is redundant in both cases

**Comment by mmarinovic** [comment]
Hi @armandooj
Any plans to merge this one?

**Comment by luciid3** [comment]
Will this be merged? 🙄

**Comment by joonarahko** [comment]
🤔 Any updates on this?

**Comment by keremciu** [comment]
yep this is needed

**Comment by javierguzman** [comment]
I am also interested on this!

**Comment by Burize** [comment]
@armandooj Hi, please fix discussions (((

**Comment by lp-bayramali** [comment]
lets merge it, it looks fine

**Comment by jemilox** [comment]
Please merge!

**Comment by AndrewAllison** [comment]
Please, please, please can we get this merged.  I might have to rethink using Formik if not.  If you need a hand at all testing or anything I'm willing to help where I can to get this sorted. @jaredpalmer

**Comment by peacheym** [comment]
Is this going to be merged?  If not, I will have to use a different form package which would be a true shame!

**Comment by mdavid626** [comment]
Can you please merge this? Everyone is waiting!

**Comment by LJunghansCode** [comment]
+1 on this merge! Let me know if I can help in any way

**Comment by lid3rs** [comment]
Any workaround on this?

**Comment by adamerose** [comment]
Can this please be merged? @jaredpalmer

**Comment by rafa8626** [comment]
@jaredpalmer any plans to merge this?

**Comment by jaredpalmer** [comment]
This is for v1?

**Comment by lid3rs** [comment]
Was it repaired?

**Comment by joonarahko** [comment]
No, was not fixed. Seeing these errors on v2 too.

**Comment by Bevilacqua** [comment]
Still seeing this issue in `2.1.4`. Is there a reason this was closed?

**Comment by gottfired** [comment]
Also having this issue. Any ETA on a fix?

**Comment by rafa8626** [comment]
@jaredpalmer no we are seeing this in v2.1.4; any chances this can get sorted out soon?

**Comment by syriley** [comment]
I am also seeing this issue. What is the plan for now? Just ignore the error?

**Comment by a-soni** [comment]
Hi, What's the plan for this?

**Comment by masakre** [comment]
also getting that warning, that merge would be appreciated

**Pull request closed without merging**